### PR TITLE
Add writer option to Client

### PR DESCRIPTION
**Pull request by chezou**
Currently, we have no way to set `Writer` with initialization for the `Client` class. This PR adds writer option to Client.

Confirmed with the following sequence as well as with adding unittest.

```py
In [1]: import pytd

In [2]: import td_pyspark

In [3]: jar_path = td_pyspark.TDSparkContextBuilder.default_jar_path()

In [5]: import os

In [6]: writer = pytd.writer.SparkWriter(apikey=os.getenv('TD_API_KEY'), endpoint=os.getenv('TD_API_SERVER'), td_spark_path=ja
   ...: r_path)
19/07/02 15:01:19 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).

In [7]: client = pytd.Client(writer=writer)

In [9]: import pandas as pd

In [10]: df = pd.DataFrame({'a':[1, 2], 'b':[3, 4]})

In [12]: client.load_table_from_dataframe(df, 'aki.test_pytd', if_exists='overwrite')
2019-07-02T15:02:29.491+0900 debug [spark] Loading com.treasuredata.spark package - (package.scala:23)
2019-07-02T15:02:29.558+0900  info [spark] td-spark version:19.7.0+1-6cc7a4ac, revision:6cc7a4a, build_time:2019-07-02T11:47:51.206+0900 - (package.scala:24)
2019-07-02T15:02:29.666+0900  info [TDServiceConfig] td-spark site: eu01 - (TDServiceConfig.scala:36)
2019-07-02T15:02:31.952+0900  info [LifeCycleManager] [session:49547178] Starting a new lifecycle ... - (LifeCycleManager.scala:187)
2019-07-02T15:02:31.957+0900  info [LifeCycleManager] [session:49547178] ======== STARTED ======== - (LifeCycleManager.scala:191)
2019-07-02T15:02:34.420+0900  warn [DefaultSource] Dropping aki.test_pytd (Overwrite mode) - (DefaultSource.scala:94)
2019-07-02T15:02:35.039+0900  info [TDWriter] Uploading data to aki.test_pytd (mode: Overwrite) - (TDWriter.scala:66)
2019-07-02T15:02:36.679+0900  info [TDWriter] [txx:b6fcb8a2] Starting a new transaction for updating aki.test_pytd - (TDWriter.scala:95)
2019-07-02T15:02:43.447+0900  info [TDWriter] [txx:b6fcb8a2] Finished uploading 1 partitions (2 records, size:127B) to aki.test_pytd - (TDWriter.scala:132)
```

### CODE REVIEW

File: pytd/client.py [hunk 1]
```diff
@@ -35,9 +35,15 @@ class Client(object):
...
+
+    writer : pytd.writer.Writer or child class of Writer class, optional
+        A Writer to choose writing method to Treasure Data. If not given, default Writer
+        will be created with executing :func:`~pytd.Client.load_table_from_dataframe`
+        at the first time.
     """
 
-    def __init__(self, apikey=None, endpoint=None, database='sample_datasets', engine='presto', header=True, **kwargs):
+    def __init__(
+            self, apikey=None, endpoint=None, database='sample_datasets', engine='presto', header=True, writer=None, **kwargs):
```
Comment (chezou) [inline pytd/client.py]: I'd love to avoid formatting discussion, so it'd be nice if we could have another PR with adding black.

Comment (yyuu) [reply pytd/client.py]: possible bikeshedding - it might be worth considering to setup [pre-commit](https://pre-commit.com/) and its [yapf plugin](https://github.com/pre-commit/mirrors-yapf) so that formatting style will be some automated task, not something derived from individual's preference :smile:

Comment (takuti) [reply pytd/client.py]: Noted 📝

File: .gitignore [hunk 2]
```diff
@@ -3,3 +3,7 @@ td-spark*
 build/
 dist/
 *.egg-info/
+__pycache__
+
+.idea
```
Comment (takuti) [inline .gitignore]: Use `.gitignore_global` for IDE/text editor specific files.

File: pytd/client.py [hunk 3]
```diff
@@ -57,7 +63,7 @@ def __init__(self, apikey=None, endpoint=None, database='sample_datasets', engin
 
         self.api_client = tdclient.Client(apikey=apikey, endpoint=endpoint, user_agent=engine.user_agent, **kwargs)
 
-        self.writer = None
+        self.writer = writer
```
Comment (takuti) [inline pytd/client.py]: I need your opinion here.

This way, non-`None` `writer` instance created outside of `Client` will be closed at the time of `Client#close`. Do you think if it's an okay behavior? 

```py
writer = Writer()
client = Client(writer=writer)
client.close()
writer.write_dataframe(...)  # failed since it's already closed inside of client
```

Guessing...users may want to reuse `writer` instance even if `client` is closed. Same discussion should happen for the `engine` argument.

File: .gitignore (hunk 2 above)
Comment (chezou) [reply .gitignore]: removed d8d7541

File: pytd/client.py (hunk 3 above)
Comment (chezou) [reply pytd/client.py]: ah, good point. 

There is a use case I can imagine that a writer could be reused; when a user creates a writer then the user uses it for both `Client` and `pandas_td`. In the case of that scenario, an engine also can be reused, but IMO the cost of creating `SparkWriter` is heavy rather than other Writes. It depends on who should manage `Writer`,  but this approach might introduce some complexity.

If you remind other use cases, I'd love to hear.

Options that I come up with are:
1) Keep it as is. A Writer will be closed and we can't reuse it.
2) Distinguish writer is created by Client or not, and close if the Writer was created by Client itself. This can be a little bit complex but we can reduce the cost of creation of SparkWriter.
3) Add some arguments something like `writer_options` which could pass extra option like `td_spark_path` (or `writer_options` can be more direct option; `td_spark_path`). This option should accept the cost of creation of SparkWriter.

I'm OK with option 3), which might be consistent with a future design change.

Comment (chezou) [reply pytd/client.py]: I found I missed applying this change to `pandas_td`. After we dicide the design, I'll work on it.

Comment (chezou) [reply pytd/client.py]: We will choose option 2, which is implemented in 69438a9 2dbc9f9

File: pytd/client.py [hunk 4]
```diff
@@ -63,6 +63,7 @@ def __init__(
 
         self.api_client = tdclient.Client(apikey=apikey, endpoint=endpoint, user_agent=engine.user_agent, **kwargs)
 
+        self.managed_writer = False
```
Comment (takuti) [inline pytd/client.py]:
```py
self.managed_writer = writer is None
```

File: pytd/client.py [hunk 5]
```diff
@@ -161,6 +162,7 @@ def load_table_from_dataframe(self, dataframe, table, if_exists='error'):
             What happens when a target table already exists.
         """
         if self.writer is None:
+            self.managed_writer = True
```
Comment (takuti) [inline pytd/client.py]: You can delete this line when the flag is initialized by `writer is None` in `__init__`

File: pytd/client.py [hunk 6]
```diff
@@ -119,7 +120,7 @@ def close(self):
         """
         self.engine.close()
         self.api_client.close()
-        if self.writer is not None:
+        if self.writer is not None and self.managed_writer:
```
Comment (takuti) [inline pytd/client.py]: Since `self.writer is not None` is always `True` where `self.managed_writer` is `False`, I personally prefer following order:

```py
if self.managed_writer and self.writer is not None:
```

For unmanaged writer, the `if` statement quickly realizes the condition becomes `False` by only checking the first sub-condition, `self.managed_writer`; the second comparison `self.writer is not None` can be skipped in the suggested ordering.

File: pytd/pandas_td/__init__.py [hunk 7]
```diff
@@ -345,8 +345,9 @@ def to_td(frame, name, con, if_exists='fail', time_col=None, time_index=None, in
     frame = _convert_index_column(frame, index, index_label)
     frame = _convert_date_format(frame, date_format)
 
-    writer = SparkWriter(con.apikey, con.endpoint)
-    writer.write_dataframe(frame, con.database, name, mode)
+    if con.writer is None:
```
Comment (takuti) [inline pytd/pandas_td/__init__.py]: Good catch - we can use `con` for writing data. 

Can it be simplified as follows?

```py
con.load_table_from_dataframe(frame, name, mode)
```

File: pytd/client.py (hunk 6 above)
Comment (chezou) [reply pytd/client.py]: Agree. fixed 80f36ae

File: pytd/pandas_td/__init__.py (hunk 7 above)
Comment (chezou) [reply pytd/pandas_td/__init__.py]: Ah, that's smarter than what I wrote 80f36ae

File: pytd/client.py (hunk 4 above)
Comment (chezou) [reply pytd/client.py]: applied 80f36ae

File: pytd/client.py (hunk 5 above)
Comment (chezou) [reply pytd/client.py]: ditto

**Comment by chezou** [comment]
hmm, CI failed. let me check
